File: android/app/src/main/kotlin/com/anhquan/unisync/features/DeviceConnection.kt
package com.anhquan.unisync.features

import com.anhquan.unisync.models.DeviceInfo
import com.anhquan.unisync.models.DeviceMessage
import com.anhquan.unisync.plugins.SocketPlugin
import com.anhquan.unisync.plugins.SocketPlugin.ConnectionState.STATE_CONNECTED
import com.anhquan.unisync.plugins.SocketPlugin.ConnectionState.STATE_DISCONNECTED
import com.anhquan.unisync.utils.ConfigUtil
import com.anhquan.unisync.utils.debugLog
import com.anhquan.unisync.utils.fromJson
import com.anhquan.unisync.utils.listen
import com.anhquan.unisync.utils.toJson
import com.anhquan.unisync.utils.warningLog
import io.reactivex.rxjava3.subjects.ReplaySubject
import javax.crypto.SecretKey

class DeviceConnection private constructor(private val socket: SocketPlugin.SocketConnection) {
    enum class DeviceState {
        ONLINE, OFFLINE, PAIRING, PAIRED
    }

    data class ConnectionNotifierValue(
        val state: DeviceState,
        val deviceInfo: DeviceInfo,
    )

    companion object {
        private val connections = mutableMapOf<String, DeviceConnection>()
        val connectionNotifier = ReplaySubject.create<ConnectionNotifierValue>()

        fun getConnections(): List<DeviceConnection> {
            return connections.values.toList()
        }

        fun getConnection(id: String): DeviceConnection {
            return connections[id]!!
        }

        fun createConnection(socket: SocketPlugin.SocketConnection) {
            DeviceConnection(socket)
        }
    }

    init {
        socket.connect()
        socket.connectionState.listen {
            when (it) {
                STATE_CONNECTED -> {
                    socket.send(toJson(ConfigUtil.Device.getDeviceInfo()))
                    socket.inputStream.listen(onNext = ::onInputData)
                }

                STATE_DISCONNECTED -> {
                    if (::info.isInitialized) {
                        onDeviceOffline()
                    }
                }

                else -> {}
            }
        }
    }

    val messageNotifier = ReplaySubject.create<DeviceMessage>()
    lateinit var info: DeviceInfo
        private set

    lateinit var secretKey: SecretKey

    private var isOnline: Boolean = false

    fun sendMessage(message: DeviceMessage) {
        if (isOnline) socket.send(toJson(message))
    }

    private fun onInputData(input: String) {
        if (!this::info.isInitialized) {
            try {
                info = fromJson(input, DeviceInfo::class.java)!!.copy(
                    ip = socket.address
                )
                onDeviceOnline()
            } catch (e: Exception) {
                socket.disconnect()
            }
        } else {
            try {
                messageNotifier.onNext(fromJson(input, DeviceMessage::class.java)!!)
            } catch (_: Exception) {
                warningLog("${this::class.simpleName}: Invalid message from ${info.name}. Message is '$input'.")
            }
        }
    }

    private fun onDeviceOnline() {
        if (connections.containsKey(info.id)) {
            debugLog("${this::class.simpleName}: Found duplicate connection to device: ${info.name} (${info.ip})")
            socket.disconnect()
        } else {
            connections[info.id] = this
            connectionNotifier.onNext(
                ConnectionNotifierValue(
                    DeviceState.ONLINE,
                    info
                )
            )
            isOnline = true
        }
    }

    private fun onDeviceOffline() {
        connections.remove(info.id)
        connectionNotifier.onNext(
            ConnectionNotifierValue(
                DeviceState.OFFLINE,
                info
            )
        )
        isOnline = false
    }
}

File: lib/backend/device_connection/device_connection.dart
// ignore_for_file: sort_constructors_first, constant_identifier_names

import 'dart:async';
import 'dart:convert';

import 'package:rxdart/rxdart.dart';
import 'package:unisync/backend/socket/socket.dart';
import 'package:unisync/models/device_info/device_info.model.dart';
import 'package:unisync/utils/configs.dart';
import 'package:unisync/utils/extensions/map.ext.dart';
import 'package:unisync/utils/logger.dart';

class DeviceConnection {
  static final Map<String, DeviceConnection> _connections = {};

  static List<DeviceConnection> get connections => _connections.values.toList();

  static void createConnection(SocketConnection socket) {
    DeviceConnection._(socket);
  }

  // ignore: close_sinks
  static final ReplaySubject<ConnectionNotifierValue> connectionNotifier = ReplaySubject();

  DeviceConnection._(this._socket) {
    initialize();
  }

  Future<void> initialize() async {
    _socket.getConnectionState().listen((event) async {
      switch (event) {
        case SocketConnectionState.STATE_CONNECTED:
          _socket.send((await ConfigUtil.device.getDeviceInfo()).toJson().toJsonString());
          _socket.getInputStream().listen(_onInputData);
          break;
        case SocketConnectionState.STATE_DISCONNECTED:
          if (info != null) {
            _onDeviceOffline();
          }
          break;
      }
    });
  }

  final SocketConnection _socket;
  // ignore: close_sinks
  final ReplaySubject<String> messageNotifier = ReplaySubject();
  DeviceInfo? info;

  void _onInputData(String input) {
    if (info == null) {
      try {
        info = DeviceInfo.fromJson(jsonDecode(input)).copy(ip: _socket.address);
        _onDeviceOnline();
      } catch (_) {
        _socket.disconnect();
      }
    } else {
      messageNotifier.add(input);
    }
  }

  void _onDeviceOnline() {
    if (_connections.containsKey(info!.id)) {
      debugLog('Found duplicate connection to device: ${info!.name} (${info!.ip})');
      _socket.disconnect();
    } else {
      _connections[info!.id] = this;
      connectionNotifier.add(ConnectionNotifierValue(DeviceState.ONLINE, info!));
    }
  }

  void _onDeviceOffline() {
    _connections.remove(info!.id);
    connectionNotifier.add(ConnectionNotifierValue(DeviceState.OFFLINE, info!));
  }
}

enum DeviceState { ONLINE, OFFLINE, MESSAGE_RECEIVED, MESSAGE_SENT }

class ConnectionNotifierValue {
  ConnectionNotifierValue(this.state, this.device);

  final DeviceState state;
  final DeviceInfo device;
}
